metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py

```diff
@@ -613,3 +648,143 @@ def get_lineage(
     def report_status(self, step: str, status: bool) -> None:
         if self.redundant_run_skip_handler:
             self.redundant_run_skip_handler.report_current_run_status(step, status)
+
+    @staticmethod
+    def _concat_sql_part(transaction_entry: Dict[int, str]) -> str:
+        sequence: int = 0
+
+        full_sql: str = ""
+
+        while sequence in transaction_entry:
+            full_sql = full_sql + transaction_entry[sequence]
+            sequence += 1
+
+        return full_sql
+
+    @staticmethod
+    def form_full_sql(temp_table_rows: List[TempTableRow]) -> List[TempTableRow]:
+        if len(temp_table_rows) == 0:
+            return []
+
+        temp_tables: List[TempTableRow] = []
+
+        # Arrange row as per transaction id and start time
+        sql_parts: Dict[Tuple[int, datetime], Dict[int, str]] = {}
+
+        for row in temp_table_rows:
+            if (row.transaction_id, row.start_time) not in sql_parts:
+                sql_parts[(row.transaction_id, row.start_time)] = {}
+            # Each part of sql has same transaction_id and start_time and unique sequence number
+            sql_parts[(row.transaction_id, row.start_time)][
+                row.sequence
+            ] = row.query_text
+
+        for row in temp_table_rows:
+            # filter out "create temp|temporary|#<table-name>" statements rows for the transaction
+            if not any(
+                row.query_text.lower().startswith(prefix)
+                for prefix in [
+                    RedshiftQuery.CREATE_TEMPORARY_TABLE_CLAUSE,
+                    RedshiftQuery.CREATE_TEMP_TABLE_CLAUSE,
+                    RedshiftQuery.CREATE_TEMPORARY_TABLE_HASH_CLAUSE,
+                ]
+            ):
+                continue
+
+            transaction_entry: Dict[int, str] = sql_parts[
+                (row.transaction_id, row.start_time)
+            ]
+
+            full_sql: str = RedshiftLineageExtractor._concat_sql_part(transaction_entry)
+
+            temp_tables.append(
+                TempTableRow(
+                    start_time=row.start_time,
+                    sequence=row.sequence,
+                    query_text=full_sql,
+                    transaction_id=row.transaction_id,
+                    process_identifier=row.process_identifier,
+                )
+            )
+
+        return temp_tables
+
+    def _add_permanent_datasets_recursively(
+        self,
+        db_name: str,
+        temp_table_names: List[str],
+        connection: redshift_connector.Connection,
+        permanent_lineage_datasets: List[LineageDataset],
+    ) -> None:
+
+        ddl_query: str = RedshiftQuery.temp_table_ddl_query(
```

Comment:
This should be cached from you earlier call otherwise this will a bunch of queries

Reply:
@lru_cache is set on function definition and I verified this function is getting called only one time